src/dune_rules/opam_create.ml

```diff
@@ -43,31 +43,50 @@ let default_build_command =
 ]
 |}))
   and from_2_9 =
-    lazy
-      (Opam_file.parse_value
-         (Lexbuf.from_string ~fname:"<internal>"
-            {|
-[
-  [ "dune" "subst" ] {dev}
-  [ "dune" "build" "-p" name "-j" jobs "--promote-install-files" "false"
+    let fname = "<internal>" in
+    let parse s = lazy (Opam_file.parse_value (Lexbuf.from_string ~fname s)) in
+    let subst = parse {| [ "dune" "subst" ] {dev} |} in
+    let build =
+      parse
+        {|
+  [ "dune" "build" "-p" name "-j" jobs
       "@install"
       "@runtest" {with-test}
       "@doc" {with-doc}
   ]
-  [ "dune" "install" "-p" name "--create-install-files" name ]
-]
-|}))
+|}
+    in
+    let install =
+      parse {| [ "dune" "install" "-p" name "--create-install-files" name ] |}
+    in
+    fun ~with_sites ->
+      let dumb_pos = (fname, 0, 0) in
+      if with_sites then
+        let build =
+          match Lazy.force build with
+          | OpamParserTypes.List (op, l) ->
+            OpamParserTypes.List
+              (op, l @ [ String (op, "--promote-install-files=false") ])
+          | _ -> assert false
+        in
+        OpamParserTypes.List
+          (dumb_pos, [ Lazy.force subst; build; Lazy.force install ])
+      else
+        OpamParserTypes.List (dumb_pos, [ Lazy.force subst; Lazy.force build ])
   in
   fun project ->
-    Lazy.force
-      (if Dune_project.dune_version project < (1, 11) then
-        before_1_11
-      else if Dune_project.dune_version project < (2, 7) then
-        from_1_11_before_2_7
-      else if Dune_project.dune_version project < (2, 9) then
-        from_2_7
-      else
-        from_2_9)
+    if Dune_project.dune_version project < (1, 11) then
+      Lazy.force before_1_11
+    else if Dune_project.dune_version project < (2, 7) then
+      Lazy.force from_1_11_before_2_7
+    else if Dune_project.dune_version project < (2, 9) then
+      Lazy.force from_2_7
+    else
```

Comment:
I believe there is another distinction to make on 2.9. There should another line, something like:
```
    else if Dune_project.dune_version project < (3, 0) then
      <whatever-the-2.9-release-generates>
```
Otherwise the opam file will change between 2.9 and 3.0 without changing the `(lang dune ...)`